### DISCUSSION TITLE
Typescript error while passing arguments to controller from middleware

### DISCUSSION THREAD

**Discussion by eyesight-pgs** [adonisjs/core#2527] [Help] [answered]
According to [Inject parameters #727](https://github.com/adonisjs/core/issues/727) thread we can pass arguments to controller from middleware like:

```ts
class FooMiddleware {
  handle (ctx, next) {
      ctx.myCustomParameter = 'foo'
      await next()
   }
}
```

It works.

But typescript gives following error:
Property 'myCustomParameter' does not exist on type 'HttpContextContract'.ts(2339)
![2021-04-28-190410_823x197_scrot](https://user-images.githubusercontent.com/33274599/116413013-d76e9580-a854-11eb-9455-4f3c93b63560.png)

And auto-completion does not work.

What I want is:
 - Typescript should not raise error
 - Auto-completion should work fine for passed argument

---

version info:
@adonisjs/core - 5.0.4-preview-rc-2.1
typescript - 4.1.5

**Comment by eyesight-pgs** [comment]
I got the following solution:

Creating new interface for myCustomParameter.
And do [type intersection](https://www.typescriptlang.org/docs/handbook/2/objects.html#intersection-types) with `HttpContextContract`

In middleware
```ts
export interface MyCustomParameter { // <-- see here
	myCustomParameter: string
}

export default class FooMiddlware {
	async handle(ctx: HttpContextContract & MyCustomParameter, next: () => Promise<void>) { // <-- see here
		ctx.myCustomParameter = 'foo'
		await next();
	}
}
```

In controller
```ts
import { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'
import { MyCustomParameter } from 'App/Middleware/FooMiddlware';

export default class TempController {
	async temp({ myCustomParameter }: HttpContextContract & MyCustomParameter) { // <-- see here
		console.log({myCustomParameter})
	}
}
```


But here we have to repeat `HttpContextContract & MyCustomParamete` in every controller where we want `myCustomParameter` which does not look ideal.

**Reply by thetutlage** [reply]
This is more of a Typescript limitation than the framework limitation.

**Reply by aayush123** [reply]
@eyesight-pgs  You can also just define `MyHttpContextContract = HttpContextContract & MyCustomParameter` in a separate TS declaration file and then use that in every controller instead of the Type provided by Adonis.
This way, you also make a maintainable type. If your middleware, in future, adds another parameter, you just have to change it in one file and it'll be available everywhere :)

**Reply by eyesight-pgs** [reply]
@thetutlage , @aayush123 Thanks for replies.


> @eyesight-pgs You can also just define `MyHttpContextContract = HttpContextContract & MyCustomParameter` in a separate TS declaration file and then use that in every controller instead of the Type provided by Adonis.
> This way, you also make a maintainable type. If your middleware, in future, adds another parameter, you just have to change it in one file and it'll be available everywhere :)

This seems to be solution to my problem. Thanks :)

**Comment by eyesight-pgs** [comment, marked as answer]
I got the exact answer from [Informing typescript about the custom property](https://docs.adonisjs.com/guides/context#informing-typescript-about-the-custom-property)

Created `context.ts` file inside `contracts` directory & added following code:
```ts
declare module '@ioc:Adonis/Core/HttpContext' {
	interface HttpContextContract {
		myCustomParameter: string
	}
}
```
and it works perfectly :)
